File: Data Structures & Algorithms/path-sum/submission-4.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func hasPathSum(root *TreeNode, targetSum int) bool {
	var dfs func(root *TreeNode, curSum int) bool
	dfs = func(root *TreeNode, curSum int) bool {
		// base case - якщо передаю пусте дерево
		if root == nil {
			return false
		}

		curSum += root.Val
		if root.Left == nil && root.Right == nil {
			return curSum == targetSum
		}

		return dfs(root.Left, curSum) || dfs(root.Right, curSum)
	}

	return dfs(root, 0)
}

File: Data Structures & Algorithms/level-order-traversal-of-binary-tree/submission-4.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

from collections import deque

class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        result = []

        if root is None:
            return result

        def bfs(root):
            queue = deque()
            queue.append(root)

            while queue:
                level_values = []
                queue_len = len(queue)

                for _ in range(queue_len):
                    current = queue.popleft()
                    level_values.append(current.val)

                    if current.left is not None:
                        queue.append(current.left)

                    if current.right is not None:
                        queue.append(current.right)

                result.append(level_values)

        bfs(root)
        return result